lib/_stream_readable.js

```diff
@@ -518,10 +518,12 @@ Readable.prototype.pipe = function(dest, pipeOpts) {
     src.once('end', endFn);
 
   dest.on('unpipe', onunpipe);
-  function onunpipe(readable) {
+  function onunpipe(readable, unpipeInfo) {
     debug('onunpipe');
     if (readable === src) {
-      cleanup();
+      if (unpipeInfo.counter++ === 0) {
```

Comment:
Nit: how about only incrementing inside to avoid incrementing when unnecessary? Or just use a boolean?